### ISSUE TITLE
Lag in drawing after mouse wheel scroll

### ISSUE THREAD

**Issue by astrofrog** [vispy/vispy#1182]
When rendering a Vispy scene, I've often noticed a delay between the scroll wheel movements and the Vispy zoom in/out. This is best illustrated when zooming in/out very fast. The following example was made on a Mac **with intertial scrolling disabled**. The animation starts off with me zooming in/out fast with the trackpad, and when the cursor goes out of the field of view, I'm no longer touching the trackpad:

![ezgif com-video-to-gif](https://cloud.githubusercontent.com/assets/314716/13355254/57b6816a-dc97-11e5-8dc9-76b589d95437.gif)

Again, this is without inertial scrolling set up. If I edit ``_qt.py`` and time the following line:

```python
        self._vispy_canvas.events.mouse_wheel(...
```

this seems to take between 0.01 and 0.05 seconds, which is quite a lot. If I go in ``perspective.py`` and edit the ``viewbox_mouse_event`` method, and comment out the following code:

```python
        if event.type == 'mouse_wheel':
            s = 1.1 ** - event.delta[1]
            self._scale_factor *= s
            if self._distance is not None:
                self._distance *= s
            self.view_changed()
```

which is what actually changes the field of view, the calls *still* take 0.01 to 0.05 seconds. Essentially, there is a large overhead in the calling of the event which is causing the wheelEvents to pile up and be processed with a significant delay. So in other words, the delay is not due (I think) to the actual zooming in/out, but to the overhead when processing the wheelEvents and figuring out what to do with them.

@Eric89GXL or anyone else, any idea of what this could be due to and how to speed things up?

**Comment by Eric89GXL** [comment]
Are you in Spyder or IPython, or running in a pure Python terminal? Interactive mode (-i) or not?

**Comment by astrofrog** [comment]
@Eric89GXL - I'm running this inside a custom Python application built with Qt that I'm starting from the terminal. Does that help?

**Comment by astrofrog** [comment]
@Eric89GXL - but I see the same effect if I run the ``examples/basics/scene/volume.py`` example from the command line with ``python volume.py``.

**Comment by Eric89GXL** [comment]
Ahh okay. I was thinking it might be related to the inputhook functionality: 

https://github.com/vispy/vispy/blob/master/vispy/app/application.py#L135

But that should only kick in with `python -i` or something like Spyder or IPython that might have its own event loop running.

Have you ever used `line_profiler`? You could try using it in the `events` module to see where the time is being spent.

**Comment by astrofrog** [comment]
Here are some initial results from ``line_profiler``. I ran an example where I zoomed in and out fast for 20 seconds or so, and it took in total 75 seconds until it stopped moving (that includes the 20 seconds of moving). During those 75 seconds, the following functions in ``event.py`` accounted for a reasonable fraction of time:

```python
Total time: 30.6699 s
File: /Users/tom/tmp/vispy/vispy/util/event.py
Function: __call__ at line 414

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   414                                               @profile
   415                                               def __call__(self, *args, **kwargs):
   416                                                   """ __call__(**kwargs)
   417                                                   Invoke all callbacks for this emitter.
   418                                           
   419                                                   Emit a new event object, created with the given keyword
   420                                                   arguments, which must match with the input arguments of the
   421                                                   corresponding event class. Note that the 'type' argument is
   422                                                   filled in by the emitter.
   423                                           
   424                                                   Alternatively, the emitter can also be called with an Event
   425                                                   instance as the only argument. In this case, the specified
   426                                                   Event will be used rather than generating a new one. This allows
   427                                                   customized Event instances to be emitted and also allows EventEmitters
   428                                                   to be chained by connecting one directly to another.
   429                                           
   430                                                   Note that the same Event instance is sent to all callbacks.
   431                                                   This allows some level of communication between the callbacks
   432                                                   (notably, via Event.handled) but also requires that callbacks
   433                                                   be careful not to inadvertently modify the Event.
   434                                                   """
   435                                                   # This is a VERY highly used method; must be fast!
   436    474600       446363      0.9      1.5          blocked = self._blocked
   437    474600       357895      0.8      1.2          if self._emitting:
   438                                                       raise RuntimeError('EventEmitter loop detected!')
   439                                           
   440                                                   # create / massage event as needed
   441    474600      8008836     16.9     26.1          event = self._prepare_event(*args, **kwargs)
   442                                           
   443                                                   # Add our source to the event; remove it after all callbacks have been
   444                                                   # invoked.
   445    474600      2365494      5.0      7.7          event._push_source(self.source)
   446    474600       402740      0.8      1.3          self._emitting = True
   447    474600       322860      0.7      1.1          try:
   448    474600       489819      1.0      1.6              if blocked.get(None, 0) > 0:  # this is the same as self.blocked()
   449         1            0      0.0      0.0                  return event
   450                                           
   451    474599       335726      0.7      1.1              rem = []
   452    915912       825672      0.9      2.7              for cb in self._callbacks:
   453    441313       448160      1.0      1.5                  if isinstance(cb, tuple):
   454    356443       361182      1.0      1.2                      obj = cb[0]()
   455    356443       250258      0.7      0.8                      if obj is None:
   456                                                                   rem.append(cb)
   457                                                                   continue
   458    356443       542933      1.5      1.8                      cb = getattr(obj, cb[1], None)
   459    356443       263426      0.7      0.9                      if cb is None:
   460    107143        72310      0.7      0.2                          continue
   461                                                           
   462    334170       342478      1.0      1.1                  if blocked.get(cb, 0) > 0:
   463                                                               continue
   464                                                           
   465    334170     10635124     31.8     34.7                  self._invoke_callback(cb, event)
   466    334170       455034      1.4      1.5                  if event.blocked:
   467                                                               break
   468                                                           
   469                                                       # remove callbacks to dead objects
   470    474599       418837      0.9      1.4              for cb in rem:
   471                                                           self.disconnect(cb)
   472                                                   finally:
   473    474600       387879      0.8      1.3              self._emitting = False
   474    474600      2595038      5.5      8.5              if event._pop_source() != self.source:
   475                                                           raise RuntimeError("Event source-stack mismatch.")
   476                                           
   477    474599       341837      0.7      1.1          return event

Total time: 28.2376 s
File: /Users/tom/tmp/vispy/vispy/util/event.py
Function: _invoke_callback at line 479

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   479                                               @profile
   480                                               def _invoke_callback(self, cb, event):
   481    334170       137642      0.4      0.5          try:
   482    334170     28099929     84.1     99.5              cb(event)
   483                                                   except Exception:
   484                                                       _handle_exception(self.ignore_callback_errors,
   485                                                                         self.print_callback_errors,
   486                                                                         self, cb_event=(cb, event))

Total time: 4.85352 s
File: /Users/tom/tmp/vispy/vispy/util/event.py
Function: _prepare_event at line 488

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   488                                               @profile
   489                                               def _prepare_event(self, *args, **kwargs):
   490                                                   # When emitting, this method is called to create or otherwise alter
   491                                                   # an event before it is sent to callbacks. Subclasses may extend
   492                                                   # this method to make custom modifications to the event.
   493    474600       464869      1.0      9.6          if len(args) == 1 and not kwargs and isinstance(args[0], Event):
   494    229561       121174      0.5      2.5              event = args[0]
   495                                                       # Ensure that the given event matches what we want to emit
   496    229561       165339      0.7      3.4              assert isinstance(event, self.event_class)
   497    245039       127932      0.5      2.6          elif not args:
   498    245039       250903      1.0      5.2              args = self.default_args.copy()
   499    245039       209721      0.9      4.3              args.update(kwargs)
   500    245039      3270259     13.3     67.4              event = self.event_class(**args)
   501                                                   else:
   502                                                       raise ValueError("Event emitters can be called with an Event "
   503                                                                        "instance or with keyword arguments only.")
   504    474600       243319      0.5      5.0          return event
```

Note that this suggests the ``__call__`` method was called over 6000 times per second, which seems a lot. I'm going to investigate which callbacks are being executed.

**Comment by astrofrog** [comment]
I ran another example that took 35-40 seconds to complete, and most of the time was spent in this function:

```python
Total time: 34.9041 s
File: /Users/tom/tmp/vispy/vispy/scene/canvas.py
Function: _process_mouse_event at line 322

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   322                                               @profile
   323                                               def _process_mouse_event(self, event):
   324       879         5526      6.3      0.0          prof = Profiler()  # noqa
   325       879          769      0.9      0.0          deliver_types = ['mouse_press', 'mouse_wheel']
   326       879         1229      1.4      0.0          if self._send_hover_events:
   327                                                       deliver_types += ['mouse_move']
   328                                           
   329       879          579      0.7      0.0          picked = self._mouse_handler
   330       879          505      0.6      0.0          if picked is None:
   331       844         1394      1.7      0.0              if event.type in deliver_types:
   332       825     26804760  32490.6     76.8                  picked = self.visual_at(event.pos)
   333                                           
   334                                                   # No visual to handle this event; bail out now
   335       879          829      0.9      0.0          if picked is None:
   336        19           12      0.6      0.0              return
   337                                                   
   338                                                   # Create an event to pass to the picked visual
   339       860        16100     18.7      0.0          scene_event = SceneMouseEvent(event=event, visual=picked)
   340                                                   
   341                                                   # Deliver the event
   342       860          790      0.9      0.0          if picked == self._mouse_handler:
   343                                                       # If we already have a mouse handler, then no other node may
   344                                                       # receive the event
   345        35           42      1.2      0.0              if event.type == 'mouse_release':
   346         2           12      6.0      0.0                  self._mouse_handler = None
   347        35       557467  15927.6      1.6              getattr(picked.events, event.type)(scene_event)
   348                                                   else:
   349                                                       # If we don't have a mouse handler, then pass the event through
   350                                                       # the chain of parents until a node accepts the event.
   351      3294         2045      0.6      0.0              while picked is not None:
   352      2471      7494490   3033.0     21.5                  getattr(picked.events, event.type)(scene_event)
   353      2471         3151      1.3      0.0                  if scene_event.handled:
   354         2            2      1.0      0.0                      if event.type == 'mouse_press':
   355         2            6      3.0      0.0                          self._mouse_handler = picked
   356         2            1      0.5      0.0                      break
   357      2469         3016      1.2      0.0                  if event.type in deliver_types:
   358                                                               # events that are not handled get passed to parent
   359      2469         4941      2.0      0.0                      picked = picked.parent
   360      2469         4101      1.7      0.0                      scene_event.visual = picked
   361                                                           else:
   362                                                               picked = None
   363                                                       
   364                                                   # If something in the scene handled the scene_event, then we mark
   365                                                   # the original event accordingly.
   366       860         2378      2.8      0.0          event.handled = scene_event.handled
```

**Comment by Eric89GXL** [comment]
Ahhh right, `visual_at`. I forgot about that. I think we have an issue open about this already, let me look.

**Comment by astrofrog** [comment]
For completeness, from a 20 second run:

```
Total time: 16.1583 s
File: /Users/tom/tmp/vispy/vispy/scene/canvas.py
Function: visual_at at line 368

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   368                                               @profile
   369                                               def visual_at(self, pos):
   370                                                   """Return the visual at a given position
   371                                           
   372                                                   Parameters
   373                                                   ----------
   374                                                   pos : tuple
   375                                                       The position in logical coordinates to query.
   376                                           
   377                                                   Returns
   378                                                   -------
   379                                                   visual : instance of Visual | None
   380                                                       The visual at the position, if it exists.
   381                                                   """
   382       458        11091     24.2      0.1          tr = self.transforms.get_transform('canvas', 'framebuffer')
   383       458        60204    131.4      0.4          fbpos = tr.map(pos)[:2]
   384                                           
   385       458          500      1.1      0.0          try:
   386       458          826      1.8      0.0              id_ = self._render_picking(region=(fbpos[0], fbpos[1],
   387       458     16079152  35107.3     99.5                                                 1, 1))
   388       458         6146     13.4      0.0              vis = VisualNode._visual_ids.get(id_[0, 0], None)
   389                                                   except RuntimeError:
   390                                                       # Don't have read_pixels() support for IPython. Fall back to
   391                                                       # bounds checking.
   392                                                       return self._visual_bounds_at(pos)
   393       458          370      0.8      0.0          return vis
```

**Comment by astrofrog** [comment]
And from a 25 second run:

```python
Total time: 16.0537 s
File: /Users/tom/tmp/vispy/vispy/scene/canvas.py
Function: _render_picking at line 455

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   455                                               @profile
   456                                               def _render_picking(self, **kwargs):
   457                                                   """Render the scene in picking mode, returning a 2D array of visual 
   458                                                   IDs.
   459                                                   """
   460       489          461      0.9      0.0          try:
   461       489       629379   1287.1      3.9              self._scene.picking = True
   462       489     14843170  30354.1     92.5              img = self.render(bgcolor=(0, 0, 0, 0), **kwargs)
   463                                                   finally:
   464       489       561797   1148.9      3.5              self._scene.picking = False
   465       489        10902     22.3      0.1          img = img.astype('int32') * [2**0, 2**8, 2**16, 2**24]
   466       489         7487     15.3      0.0          id_ = img.sum(axis=2).astype('int32')
   467       489          469      1.0      0.0          return id_

Total time: 14.8183 s
File: /Users/tom/tmp/vispy/vispy/scene/canvas.py
Function: render at line 209

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   209                                               @profile
   210                                               def render(self, region=None, size=None, bgcolor=None):
   211                                                   """Render the scene to an offscreen buffer and return the image array.
   212                                                   
   213                                                   Parameters
   214                                                   ----------
   215                                                   region : tuple | None
   216                                                       Specifies the region of the canvas to render. Format is 
   217                                                       (x, y, w, h). By default, the entire canvas is rendered.
   218                                                   size : tuple | None
   219                                                       Specifies the size of the image array to return. If no size is 
   220                                                       given, then the size of the *region* is used, multiplied by the 
   221                                                       pixel scaling factor of the canvas (see `pixel_scale`). This
   222                                                       argument allows the scene to be rendered at resolutions different
   223                                                       from the native canvas resolution.
   224                                                   bgcolor : instance of Color | None
   225                                                       The background color to use.
   226                                           
   227                                                   Returns
   228                                                   -------
   229                                                   image : array
   230                                                       Numpy array of type ubyte and shape (h, w, 4). Index [0, 0] is the 
   231                                                       upper-left corner of the rendered region.
   232                                                   
   233                                                   """
   234       489         7506     15.3      0.1          self.set_current()
   235                                                   # Set up a framebuffer to render to
   236       489          877      1.8      0.0          offset = (0, 0) if region is None else region[:2]
   237       489          591      1.2      0.0          csize = self.size if region is None else region[2:]
   238       489         7764     15.9      0.1          s = self.pixel_scale
   239       489         2421      5.0      0.0          size = tuple([x * s for x in csize]) if size is None else size
   240       489        20679     42.3      0.1          fbo = gloo.FrameBuffer(color=gloo.RenderBuffer(size[::-1]),
   241       489        69681    142.5      0.5                                 depth=gloo.RenderBuffer(size[::-1]))
   242                                           
   243       489      4774976   9764.8     32.2          self.push_fbo(fbo, offset, csize)
   244       489          667      1.4      0.0          try:
   245       489       773362   1581.5      5.2              self._draw_scene(bgcolor=bgcolor)
   246       489      2785959   5697.3     18.8              return fbo.read()
   247                                                   finally:
   248       489      6373825  13034.4     43.0              self.pop_fbo()
```

So most of the time is spent rendering the visualization to offscreen buffers inside ``visual_at``.

**Comment by Eric89GXL** [comment]
I guess I never opened an issue, just mentioned it:

https://github.com/vispy/vispy/pull/1021#issuecomment-125779617

TLDR; Right now all mouse interactions require rendering the scene to framebuffer to send events, but for many applications we could use the position and some logic to get e.g. scrolling behavior.

Can you change the title to something like "ENH: Allow mouse-position-based and picking-based interaction in scene"?

**Comment by Eric89GXL** [comment]
Or really this comment https://github.com/vispy/vispy/pull/1021#issuecomment-125773989

**Comment by astrofrog** [comment]
@Eric89GXL - just to make sure I understand, what are possible cases where you don't want the scrolling to just be passed to the ViewBox?

**Comment by Eric89GXL** [comment]
I think that there should be two mouse-interaction modes:

1. Picking-based (what we have now)

2. Widget- and mouse-position-based

The former is useful when you want to be able to pick individual items from a plot, and also works for zooming and panning, but is *much* slower for those cases. The latter is useful when you just want to zoom in and out, or maybe pan and zoom, but can't pick individual items from within a widget (like ViewBox). The latter would work by using the known layout of scenegraph elements (viewboxes, widgets), and passing the event to the correct widget based solely on mouse position and the x,y,w,h of those elements. The widget/mouse-based one is actually what we used to have, but we replaced it with the more general but slower picking-based method.

**Comment by astrofrog** [comment]
@Eric89GXL - ok, thanks for clarifying this! I like the idea of having two mouse modes - in the particular case I'm looking at, we actually never need to pick items from a plot, so it would be nice to be able to have the faster algorithm for that.

I think this would be best implemented by someone more familiar with the code than me though, since it may be possible to re-use parts of the old implementation to save time?

**Comment by Eric89GXL** [comment]
I think at this point anybody other than maybe @campagnola would have to work a bit to figure out where it would go :) Feel free to take a stab at it if you have time. I don't expect it to be trivial, though, unfortunately.

**Comment by Eric89GXL** [comment]
@astrofrog any way you want to tackle this monster for 0.5? :)

**Comment by campagnola** [comment]
FWIW: visuals have an `interactive` flag. If you disable that flag, then they are not rendered for picking. Is it possible that your volume is being rendered into the picking buffer?

**Comment by codypiersall** [comment]
I've been looking in to #1274 still (closed as a duplicate of this one) and I found that for some reason, adding padding in the `PlotWidget` causes the plots to become much less responsive.  I haven't figured out why yet, but I'm still looking in to it.  The following patch which removes padding causes the mouse interactivity to behave much better:

```
diff --git a/vispy/plot/plotwidget.py b/vispy/plot/plotwidget.py
index de6fcef..25cc56a 100644
--- a/vispy/plot/plotwidget.py
+++ b/vispy/plot/plotwidget.py
@@ -77,29 +77,14 @@ class PlotWidget(scene.Widget):
         #     |                           padding                           |
         #     +---------+-----------------------+-------+---------+---------+
 
-        # padding left
-        padding_left = self.grid.add_widget(None, row=0, row_span=5, col=0)
-        padding_left.width_min = 30
-        padding_left.width_max = 60
-
-        # padding right
-        padding_right = self.grid.add_widget(None, row=0, row_span=5, col=6)
-        padding_right.width_min = 30
-        padding_right.width_max = 60
-
-        # padding right
-        padding_bottom = self.grid.add_widget(None, row=6, col=0, col_span=6)
-        padding_bottom.height_min = 20
-        padding_bottom.height_max = 40
-
         # row 0
         # title - column 4 to 5
-        self.title_widget = self.grid.add_widget(self.title, row=0, col=4)
+        self.title_widget = self.grid.add_widget(self.title, row=0, col=3)
         self.title_widget.height_min = self.title_widget.height_max = 40
 
         # row 1
         # colorbar - column 4 to 5
-        self.cbar_top = self.grid.add_widget(None, row=1, col=4)
+        self.cbar_top = self.grid.add_widget(None, row=1, col=3)
         self.cbar_top.height_max = 1
 
         # row 2
@@ -108,43 +93,43 @@ class PlotWidget(scene.Widget):
         # yaxis - column 3
         # view - column 4
         # colorbar_right - column 5
-        self.cbar_left = self.grid.add_widget(None, row=2, col=1)
+        self.cbar_left = self.grid.add_widget(None, row=2, col=0)
         self.cbar_left.width_max = 1
 
         self.ylabel = scene.Label("", rotation=-90)
-        ylabel_widget = self.grid.add_widget(self.ylabel, row=2, col=2)
+        ylabel_widget = self.grid.add_widget(self.ylabel, row=2, col=1)
         ylabel_widget.width_max = 1
 
         self.yaxis = scene.AxisWidget(orientation='left',
                                       text_color=fg,
                                       axis_color=fg, tick_color=fg)
 
-        yaxis_widget = self.grid.add_widget(self.yaxis, row=2, col=3)
+        yaxis_widget = self.grid.add_widget(self.yaxis, row=2, col=2)
         yaxis_widget.width_max = 40
 
-        self.view = self.grid.add_view(row=2, col=4,
+        self.view = self.grid.add_view(row=2, col=3,
                                        border_color='grey', bgcolor="#efefef")
         self.view.camera = 'panzoom'
         self.camera = self.view.camera
 
-        self.cbar_right = self.grid.add_widget(None, row=2, col=5)
+        self.cbar_right = self.grid.add_widget(None, row=2, col=4)
         self.cbar_right.width_max = 1
 
         # row 3
         # xaxis - column 4
         self.xaxis = scene.AxisWidget(orientation='bottom', text_color=fg,
                                       axis_color=fg, tick_color=fg)
-        xaxis_widget = self.grid.add_widget(self.xaxis, row=3, col=4)
+        xaxis_widget = self.grid.add_widget(self.xaxis, row=3, col=3)
         xaxis_widget.height_max = 40
 
         # row 4
         # xlabel - column 4
         self.xlabel = scene.Label("")
-        xlabel_widget = self.grid.add_widget(self.xlabel, row=4, col=4)
+        xlabel_widget = self.grid.add_widget(self.xlabel, row=4, col=3)
         xlabel_widget.height_max = 40
 
         # row 5
-        self.cbar_bottom = self.grid.add_widget(None, row=5, col=4)
+        self.cbar_bottom = self.grid.add_widget(None, row=5, col=3)
         self.cbar_bottom.height_max = 1
 
         self._configured = True
```

And here are two videos that display the differences:

## Without patch
![bad-movement](https://cloud.githubusercontent.com/assets/4684327/19256910/d9931ecc-8f30-11e6-8bea-4cf4616c6598.gif)

## With patch
![good-movement](https://cloud.githubusercontent.com/assets/4684327/19256927/204e4f44-8f31-11e6-9a34-7b1c2073bea5.gif)

------------------

I'm not suggesting that removing the padding is actually a good solution, but I figured that it might give you guys more information who actually understand how VisPy works.

**Comment by codypiersall** [comment]
It turns out that **the slowness/lagginess reported in this issue is actually due to time spent using the Cassowary solver, as mentioned in #1049**, I believe.  After running Python's [cProfile](https://docs.python.org/3.5/library/profile.html#module-cProfile) on this plot:

    python -m cProfile -o volume.prof  examples/basics/plotting/volume.py

And scrolling around for about a minute, then examining with [pstats](https://docs.python.org/3.5/library/profile.html#module-pstats):

```
python -m pstats volume.prof
(horusui) C:\Users\pier3595\Dev\vispy>python -m pstats volume.prof
Welcome to the profile statistics browser.
volume.prof% sort tottime
volume.prof% stats 10
Tue Oct 11 14:50:51 2016    volume.prof

         223825072 function calls (223201272 primitive calls) in 100.654 seconds

   Ordered by: internal time
   List reduced from 3564 to 10 due to restriction <10>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
  3333419   34.156    0.000   71.231    0.000 C:\Users\pier3595\Dev\vispy\vispy\ext\_bundled\cassowary\expression.py:375(substitute_out)
 18370756   10.661    0.000   14.103    0.000 C:\Users\pier3595\Dev\vispy\vispy\ext\_bundled\cassowary\tableau.py:35(note_added_variable)
 22324182    6.785    0.000    6.785    0.000 C:\Users\pier3595\Dev\vispy\vispy\ext\_bundled\cassowary\expression.py:357(set_variable)
    53709    5.890    0.000    7.673    0.000 {built-in method builtins.sorted}
 18437826    4.938    0.000    6.090    0.000 C:\Users\pier3595\Dev\vispy\vispy\ext\_bundled\cassowary\utils.py:12(approx_equal)
 15031541    4.578    0.000    5.697    0.000 C:\Users\pier3595\Dev\vispy\vispy\ext\_bundled\cassowary\tableau.py:31(note_removed_variable)
 40015639    4.320    0.000    4.320    0.000 {method 'get' of 'dict' objects}
     5244    2.992    0.001   85.249    0.016 C:\Users\pier3595\Dev\vispy\vispy\ext\_bundled\cassowary\simplex_solver.py:512(optimize)
    48385    2.102    0.000   73.333    0.002 C:\Users\pier3595\Dev\vispy\vispy\ext\_bundled\cassowary\tableau.py:94(substitute_out)
 19017321    1.891    0.000    1.891    0.000 {method 'add' of 'set' objects}
```

You can see that **over a third of the time is spent in `substitute_out`**, and **over 85% of the cumulative time is spent in `optimize()`**

I also [uploaded a call graph](http://arrc.ou.edu/~cody/volume_call_graph.svg) generated from running this example that you can check out too if you want.

I'm still trying to really get to the bottom of this, but now that I know where the problem is coming from I'm more hopeful.  Any helpful hints would be appreciated.

**Comment by Eric89GXL** [comment]
Argh. That deserves a new issue, since it's orthogonal to the slowdowns caused by this one. Can you open one about Cassowary being a bottleneck?

Does this happen on resizing, or even during zooming? Cassowary shouldn't need to be called when simply zooming or rotating a plot, only when widgets get resized.

**Comment by codypiersall** [comment]
The issue is in `vispy/scene/widgets/grid/_update_child_widget_dim()`.  The problem is that basically _all the constraints are updated every time any child widget needs to be updated_ because the `_height_stay` and `_width_stay` are removed and added whether they need to be or not.  I'm working on figuring out the best way to only conditionally update them.  I'm pretty sure their purpose is so that resizing the window causes the grid to properly resize.  This is around line 463 of grid.py.  I'll work on a PR to fix the issue.  Hopefully I'll have a PR tomorrow.

**Comment by Eric89GXL** [comment]
> I'm working on figuring out the best way to only conditionally update them. I'm pretty sure their purpose is so that resizing the window causes the grid to properly resize. This is around line 463 of grid.py. I'll work on a PR to fix the issue.

Great! That is much better than opening a new issue :) Sounds like the right approach, I agree there should be no need to update those every time.

**Comment by codypiersall** [comment]
I [created a branch](https://github.com/codypiersall/vispy/tree/cassowary-perf) that _sort of_ fixes the problem, but I'm not convinced I've done it the best way.  Mouse events are super snappy, but resize events are still super slow.  I'll work on it a bit more tomorrow and see if I can solve the resize problem as well.

**Comment by Eric89GXL** [comment]
I think it's okay if resize is slower, those should be much rarer than
mouse events
